File: problems/15/solution.py
from collections import Counter, defaultdict


def can_construct_no_counter(letter: str, magazine: str) -> bool:
    magazine_count = defaultdict(int)
    for ch in magazine:
        magazine_count[ch] += 1

    letter_count = defaultdict(int)
    for ch in letter:
        letter_count[ch] += 1

    for ch, cnt in letter_count.items():
        if cnt > magazine_count[ch]:
            return False
    return True


def can_construct(letter: str, magazine: str) -> bool:
    magazine_count = Counter(magazine)
    letter_count = Counter(letter)
    for ch, cnt in letter_count.items():
        if cnt > magazine_count[ch]:
            return False
    return True


def test_can_construct():
    assert can_construct("a", "b") == False
    assert can_construct("aa", "ab") == False
    assert can_construct("aa", "aab") == True

    assert can_construct_no_counter("a", "b") == False
    assert can_construct_no_counter("aa", "ab") == False
    assert can_construct_no_counter("aa", "aab") == True


if __name__ == "__main__":
    import pytest
    pytest.main(["-xvv", __file__])

File: problems/15/solution.go
package main

type LetterCount [26]int

func CanConstruct(letter string, magazine string) bool {
	var letterCount LetterCount
	var magazineCount LetterCount

	for _, ch := range letter {
		index := int(byte(ch) - 'a')
		letterCount[index]++
	}

	for _, ch := range magazine {
		index := int(byte(ch) - 'a')
		magazineCount[index]++
	}

	for ch, cnt := range letterCount {
		if cnt > magazineCount[ch] {
			return false
		}
	}

	return true
}
